pkg/pillar/cmd/zedagent/zedagent.go

```diff
@@ -291,323 +435,764 @@ func Run(ps *pubsub.PubSub, loggerArg *logrus.Logger, logArg *base.LogObject) in
+	zedagentCtx.cipherCtx = cipherCtx
 
-	attestCtx.zedagentCtx = &zedagentCtx
-	zedagentCtx.attestCtx = &attestCtx
+	attestCtx.zedagentCtx = zedagentCtx
+	zedagentCtx.attestCtx = attestCtx
+	return zedagentCtx
+}
 
-	pubZedAgentStatus, err := ps.NewPublication(pubsub.PublicationOptions{
-		AgentName: agentName,
-		TopicType: types.ZedAgentStatus{},
-	})
-	if err != nil {
-		log.Fatal(err)
+func initializeDirs() {
+	// create persistent holder directory
+	if _, err := os.Stat(types.PersistDir); err != nil {
+		log.Tracef("Create %s", types.PersistDir)
+		if err := os.MkdirAll(types.PersistDir, 0700); err != nil {
+			log.Fatal(err)
+		}
 	}
-	pubZedAgentStatus.ClearRestarted()
-	getconfigCtx.pubZedAgentStatus = pubZedAgentStatus
-
-	// apply saved radio config ASAP
-	initializeRadioConfig(&getconfigCtx)
-
-	// Wait until we have been onboarded aka know our own UUID
-	subOnboardStatus, err := ps.NewSubscription(pubsub.SubscriptionOptions{
-		AgentName:     "zedclient",
-		MyAgentName:   agentName,
-		TopicImpl:     types.OnboardingStatus{},
-		Activate:      true,
-		Persistent:    true,
-		Ctx:           &zedagentCtx,
-		CreateHandler: handleOnboardStatusCreate,
-		ModifyHandler: handleOnboardStatusModify,
-		WarningTime:   warningTime,
-		ErrorTime:     errorTime,
-	})
-	if err != nil {
-		log.Fatal(err)
+	if _, err := os.Stat(types.CertificateDirname); err != nil {
+		log.Tracef("Create %s", types.CertificateDirname)
+		if err := os.MkdirAll(types.CertificateDirname, 0700); err != nil {
+			log.Fatal(err)
+		}
+	}
+	if _, err := os.Stat(checkpointDirname); err != nil {
+		log.Tracef("Create %s", checkpointDirname)
+		if err := os.MkdirAll(checkpointDirname, 0700); err != nil {
+			log.Fatal(err)
+		}
 	}
-	// Wait for Onboarding to be done by client
+}
+
+func waitUntilOnboarded(zedagentCtx *zedagentContext, stillRunning *time.Ticker) {
+	zedagentCtx.subOnboardStatus.Activate()
 	nilUUID := uuid.UUID{}
 	for devUUID == nilUUID {
 		log.Functionf("Waiting for OnboardStatus UUID")
 		select {
-		case change := <-subOnboardStatus.MsgChan():
-			subOnboardStatus.ProcessChange(change)
+		case change := <-zedagentCtx.subOnboardStatus.MsgChan():
+			zedagentCtx.subOnboardStatus.ProcessChange(change)
 		case <-stillRunning.C:
 		}
-		ps.StillRunning(agentName, warningTime, errorTime)
+		zedagentCtx.ps.StillRunning(agentName, warningTime, errorTime)
 	}
+}
 
-	// We know our own UUID; prepare for communication with controller
-	zedcloudCtx = handleConfigInit(
-		zedagentCtx.globalConfig.GlobalValueInt(types.NetworkSendTimeout),
-		zedagentCtx.zedcloudMetrics)
-	// Timer for deferred sends of info messages
-	deferredChan := zedcloud.GetDeferredChan(zedcloudCtx, getDeferredSentHandlerFunction(&zedagentCtx), getDeferredPriorityFunctions()...)
+func waitUntilGCReady(zedagentCtx *zedagentContext, stillRunning *time.Ticker) {
+	getconfigCtx := zedagentCtx.getconfigCtx
+	zedagentCtx.subGlobalConfig.Activate()
+	getconfigCtx.subNodeAgentStatus.Activate()
```

Comment:
It become quite tricky to understand what we should and should not activate and where. For example here we rely on `waitUntilOnboarded` and if we will modify logic or order in another function we will hit problems here. At least it will be great to add comment that we expected another subscription activated previously. But better to check somehow and activate all required subscriptions inside the function that use them.